### PR TITLE
NIFI-9565: Added StartTextractJob, FetchTextractResults, and InvokeTe…

### PR DESCRIPTION
**Pull request by markap14**
…xtract processors

- Added StartSpeechSynthesisTask and FetchSpeechSynthesisResults

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by nandorsoma** [comment]
reviewing

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/textract/StartTextractJob.java [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+        .description("The ARN of the Amazon Role that Textract is to use in order to publish results to the configured SNS Topic")
+        .required(true)
+        .addValidator(NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(FLOWFILE_ATTRIBUTES)
+        .build();
+    static final PropertyDescriptor OUTPUT_S3_BUCKET = new PropertyDescriptor.Builder()
+        .name("Output S3 Bucket")
+        .displayName("Output S3 Bucket")
+        .description("The name of the S3 Bucket that the Textract output should be written to")
+        .required(true)
```
Comment (nandorsoma) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/textract/StartTextractJob.java:164]: I'm wondering that couldn't we make this property optional? When I tested I've seen that when Textract finished its job it saved the results to s3, but FetchTextractResults also wrote the result into the FlowFile's content. What happens if I don't want this duplication?

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/textract/FetchTextractResults.java [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+@CapabilityDescription("Retrieves the results of a Textract job that was triggered via StartTextractJob. If the job has not completed yet, the incoming FlowFile will be routed to the 'in progress' " +
+    "Relationship, providing the ability to retry after the Penalization period. Otherwise, the original FlowFile will be transferred to the 'original' relationship while the results of the " +
+    "Textract job are transferred as JSON to the 'success' or 'partial success' relationship. This processor is typically used in conjunction with GetSQS and/or StartTextractJob. See Additional " +
+    "Details for more information.")
+@SupportsBatching
+@SideEffectFree
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@Tags({"aws", "cloud", "text", "textract", "unstructured", "detect", "analyze"})
+@SeeAlso({AWSCredentialsProviderService.class, StartTextractJob.class, InvokeTextract.class})
+public class FetchTextractResults extends AbstractAWSCredentialsProviderProcessor<AmazonTextractAsyncClient> {
```
Comment (nandorsoma) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/textract/FetchTextractResults.java:89]: This processor fetches the actual result of the textract job and puts it to the flowfile. On the other hand FetchSpeechSynthesisResults just fetches the status and the actual result is fetched only with the FetchS3Object processor. Wouldn't it make sense to make the behavior of this processor similar to be consistent across these features?

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/textract/InvokeTextract.java [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+@InputRequirement(Requirement.INPUT_REQUIRED)
+@SideEffectFree
+@CapabilityDescription("Sends the contents of the FlowFile to the AWS Textract service in order to perform text-related extraction and analysis on an image. " +
+    "Upon successful completion, the original FlowFile will be updated with any relevant attributes, such as the raw extracted text. A second FlowFile will be " +
```
Comment (nandorsoma) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/textract/InvokeTextract.java:97]: The documentation of the original relationship says "Upon successful analysis, the original FlowFile will be routed to the 'original' relationship while the results of the analysis are routed to the 'analysis' relationship". I've also tested it and the original FlowFile didn't contain the extracted text, but the analyzed one yes. Could you verify that I'm right and update the description if needed?

File: nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/textract/StartTextractJob.java [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+        .build();
+    static final PropertyDescriptor KMS_KEY_ID = new PropertyDescriptor.Builder()
+        .name("KMS Key ID")
+        .displayName("KMS Key ID")
+        .description("If specified, the key whose ID is provided will be used to encrypt the results before writing the results to S3")
+        .required(false)
+        .addValidator(NON_EMPTY_VALIDATOR)
+        .expressionLanguageSupported(FLOWFILE_ATTRIBUTES)
+        .build();
+    static final PropertyDescriptor SNS_TOPIC_ARN = new PropertyDescriptor.Builder()
```
Comment (nandorsoma) [inline nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/textract/StartTextractJob.java:144]: It's not really clear to me why SNS is needed. If I understand correctly FetchTextractResults uses the textract api to fetch the results and it doesn't use the SNS topic to get notified.

**Review by nandorsoma — CHANGES_REQUESTED** [review]
Hi Mark!
Thank you for implementing these processors. I think it makes easier to use these aws provided features. I always knew that these features were there but it wasn't easy to start working with them. I wouldn't say it is easy now, but the processors and the example that you provided together make a good guide to start using them.

I've found a few bugs and I have some questions that I commented inline. Could you check them, please?

And another thing that I wasn't able to comment inline:
speech-synthesis-dataflow.png is added two times to the repository. Probably we could mention with a link in the Fetch* processors' additionalDetails that there are useful information in the Start Speech Synthesis Task's additionalDetails. This way we don't have to duplicate png files in the repository. Also, as I've tested all three scenarios, I have screenshots for the Textract sync and async flows. Do you need them so we can add them to the documentation like you did for speech synthesis?

**Comment by markap14** [comment]
After looking at some of the other AWS ML services, we will need to configure them quite differently because the configuration will be far more complex. Because of that, we may need to rethink how we configure these processors as well. As a result, I'm going to close this PR for now.

**Pull request closed without merging**